File: Spark/src/main/scala/com/tu/bdap/hashmin/HashMin_BDAP.scala
package com.tu.bdap.hashmin

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import org.apache.spark.graphx.Edge
import org.apache.spark.graphx.Graph

import com.tu.bdap.utils.DataSetLoader


/**
  * This algorithm computes connected components by propagating the lowest vertex-id to all neighbours
  */
object HashMin_BDAP {
  /**
    * Loads a graph from local disk or hdfs and calculates CC using HashMin
    * @param args args[0] should contain path, args[1] is an integer identifying the dataset
    */
  def main(args: Array[String]): Unit = {

    //Start the Spark context
    val conf = new SparkConf()
      .setAppName("HashMin")
    //.setMaster("local")

    val sc = new SparkContext(conf)

    //Set NumIterations
    val numIterations = 10

    //Check input arguments
    if (args.length < 2) {
      System.err.println("Invalid input arguments")
      return
    }

    //Set input Arguments
    val dataSetPath = args(0)
    val dataSet = args(1).toInt

    //Load DataSet
    var edges = dataSet match {
      case 1 => DataSetLoader.loadUSA(sc, dataSetPath)
      case 2 => DataSetLoader.loadTwitter(sc, dataSetPath)
      case 3 => DataSetLoader.loadFriendster(sc, dataSetPath)
      case _ => null
    }

    //Make the Graph undirected
    edges = edges.flatMap(edge => Seq(Edge(edge.srcId, edge.dstId, 0), Edge(edge.dstId, edge.srcId, 0)))

    //Check if DataSet could be loaded
    if (edges == null) {
      System.err.println("Could not load DataSet")
      return
    }

    //Create Graph from edges
    var graph = Graph.fromEdges(edges, 0.0)

    graph = graph.mapVertices((id, _) => id.toDouble)


    //1. Superstep: Send vertex-id to all neighbours
    //2. Superstep: If smallest received vertex-id is smaller then current one, update it and send to all neighbours again
    graph = graph.pregel(Double.PositiveInfinity, numIterations)(
      (id, dist, newDist) => math.min(dist, newDist), // Vertex Program
      triplet => { // Send Message
        if (triplet.srcAttr < triplet.dstAttr) {
          Iterator((triplet.dstId, triplet.srcAttr))
        } else {
          Iterator.empty
        }
      },
      (a, b) => math.min(a, b) // Merge Message
      )

    graph.vertices.collect()

  }
}

File: Flink/src/main/java/com/tu/bdap/hashmin/HashMin_BDAP.java
package com.tu.bdap.hashmin;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.DataSet;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.graph.Edge;
import org.apache.flink.graph.Graph;
import org.apache.flink.graph.Vertex;
import org.apache.flink.graph.pregel.ComputeFunction;
import org.apache.flink.graph.pregel.MessageIterator;

import com.tu.bdap.utils.DataSetID;
import com.tu.bdap.utils.DatasetLoader;


//This algorithm computes connected components by sending the lowest vertex-id to all neighbours
public class HashMin_BDAP {

	public static void main(String[] args) throws Exception {

		// Initialize DataSetPath and DataSet
		String dataSetPath = "";
		int dataSet = -1;

		try {
			dataSetPath = args[0];
			dataSet=Integer.parseInt(args[1]);
		} catch (Exception e) {
			e.printStackTrace();
		}

		if (dataSetPath == "" || dataSet == -1) {
			System.err.println("Wrong input parameters!");
			return;
		}

		// Iterations
		int numIterations = 11;

		// Create execution environment
		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

		// Initialize DataSet with GraphData
		DataSet<Edge<Integer, Double>> edges = null;

		// Read DataSet with GraphData using GraphLoader Class
		switch (dataSet) {
		case DataSetID.USA:
			edges = DatasetLoader.loadUsaInt(dataSetPath, env);
			break;
		case DataSetID.TWITTER:
			edges = DatasetLoader.loadTwitterInt(dataSetPath, env);
			break;
		case DataSetID.FRIENDSTER:
			edges = DatasetLoader.loadFriendsterInt(dataSetPath, env);
			break;
		}
		
		//Check if DataSet could be loaded!
		if(edges==null)
		{
			System.err.println("DataSet could not be loaded!");
			return;
		}

		// Create Graph from DataSet and make it undirected
		Graph<Integer, Double, Double> graph = Graph.fromDataSet(edges, new MapFunction<Integer, Double>() {
			@Override
			public Double map(Integer value) throws Exception {
				return (double) value;
			}
		}, env).getUndirected();

		// Run HashMin Algorithm on the Graph
		Graph<Integer, Double, Double> result = graph.runVertexCentricIteration(new HashMinComputeFunction(), null,
				numIterations);

		// Extract the vertices as the result
		DataSet<Vertex<Integer, Double>> hashmin = result.getVertices();

		hashmin.collect();

	}

	public static final class HashMinComputeFunction extends ComputeFunction<Integer, Double, Double, Double> {

		@Override
		public void compute(Vertex<Integer, Double> vertex, MessageIterator<Double> messages) {

			// First superstep: send vertex-id to each neighbour
			if (getSuperstepNumber() == 1) {
				for (Edge<Integer, Double> e : getEdges()) {
					sendMessageTo(e.getTarget(), vertex.getValue());
				}
				// All other supersteps:
			} else {
				// Find the smallest vertex-id which was recived
				double minMessage = Double.POSITIVE_INFINITY;
				for (Double msg : messages) {
					minMessage = Math.min(minMessage, msg);
				}
				// If the smallest vertex-id which was recived is smaller than the
				// current one, update it and send to all neighbours again
				if (minMessage < vertex.getValue()) {
					setNewVertexValue(minMessage);
					for (Edge<Integer, Double> e : getEdges()) {
						sendMessageTo(e.getTarget(), minMessage);
					}
				}

			}
		}
	}

}
